src/imageloader.js

```diff
@@ -71,93 +72,29 @@ ImageJob.prototype = {
-                headers: this.ajaxHeaders,
-                responseType: "arraybuffer",
-                postData: this.postData,
-                success: function(request) {
-                    var blb;
-                    // Make the raw data into a blob.
-                    // BlobBuilder fallback adapted from
-                    // http://stackoverflow.com/questions/15293694/blob-constructor-browser-compatibility
-                    try {
-                        blb = new window.Blob([request.response]);
-                    } catch (e) {
-                        var BlobBuilder = (
-                            window.BlobBuilder ||
-                            window.WebKitBlobBuilder ||
-                            window.MozBlobBuilder ||
-                            window.MSBlobBuilder
-                        );
-                        if (e.name === 'TypeError' && BlobBuilder) {
-                            var bb = new BlobBuilder();
-                            bb.append(request.response);
-                            blb = bb.getBlob();
-                        }
-                    }
-                    // If the blob is empty for some reason consider the image load a failure.
-                    if (blb.size === 0) {
-                        self.errorMsg = "Empty image response.";
-                        self.finish(false);
-                    }
-                    // Create a URL for the blob data and make it the source of the image object.
-                    // This will still trigger Image.onload to indicate a successful tile load.
-                    var url = (window.URL || window.webkitURL).createObjectURL(blb);
-                    self.image.src = url;
-                },
-                error: function(request) {
-                    self.errorMsg = "Image load aborted - XHR error: Ajax returned " + request.status;
-                    self.finish(false);
-                }
-            });
-
-            // Provide a function to properly abort the request.
-            this.abort = function() {
-                self.request.abort();
-
-                // Call the existing abort function if available
-                if (typeof selfAbort === "function") {
-                    selfAbort();
-                }
-            };
-        } else {
-            if (this.crossOriginPolicy !== false) {
-                this.image.crossOrigin = this.crossOriginPolicy;
+        this.abort = function() {
+            self.request.abort();
+            if (typeof selfAbort === "function") {
+                selfAbort();
             }
+        };
 
-            this.image.src = this.src;
-        }
+        this.source.downloadTileStart(this);
     },
 
-    finish: function(successful) {
-        this.image.onload = this.image.onerror = this.image.onabort = null;
-        if (!successful) {
-            this.image = null;
-        }
+    finish: function(successful, errorMessage) {
```

Comment:
The change is due to the API being exposed, the user would prefer saying `ctx.finish(false, "Something was wrong IDK") ` rather than `ctx.errMsg = "Something was wrong IDK"; ctx.finish(false, ) `. Yes, the old way on line 80 works too, but it would be better to use the finish argument, I just overlooked it since I had to combine this PR from several sources and this little part got away unchanged.

Reply:
Or we could event just say `finish()` which would be considered unsuccessful if it contained a message...

Reply:
`finish()` could also receive the data object instead of boolean, so that we could remove `downloadTileFinish` from TileSource... any preference?

Reply:
Yeah, seems like it might be a simpler API if we had `finish` get called with the data (or null) and an optional error message.